core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java

```diff
@@ -447,6 +459,21 @@ private ResponseHeaders setHttpStatus(ResponseHeadersBuilder headers) {
         return CompletableFuture.completedFuture(obj);
     }
 
+    /**
+     * Propagates {@link CompletionStage downstream} error to {@link CompletionStage upstream}.
+     */
+    private static CompletionStage<?> cancelOnDownstreamException(
+            CompletionStage<?> upstream, CompletionStage<? extends HttpResponse> downstream) {
+        downstream.exceptionally(cause -> {
+            final CompletableFuture<?> upstreamFuture = upstream.toCompletableFuture();
+            if (!upstreamFuture.isDone()) {
+                upstreamFuture.completeExceptionally(cause);
+            }
+            return null;
+        });
+        return upstream;
```

Comment:
Question: should we remove https://github.com/line/armeria/pull/3691/files#diff-376b1094f9e3a90b5d82ef6a3799eb9e07bf0cf2a5b6e5921da9b10ce1c2154bR89-R94 in `ResponseConversionUtil.java` and https://github.com/line/armeria/pull/3691/files#diff-bd13fd7b1fbdc0f156dc2031491affde55933cb442ac1d2ab3480ac081fa643aR111 in `ObjectCollectingUtil.java` 
if we add a callback for `HttpResponse` here?
```suggestion
        downstream.handle((httpResponse, cause) -> {
            if (cause != null) {
	            final CompletableFuture<?> upstreamFuture = upstream.toCompletableFuture();
	            if (!upstreamFuture.isDone()) {
	                upstreamFuture.completeExceptionally(cause);
	            }
	            return null;
            }
            
            httpResponse.whenComplete().exceptionally(cause -> {
               // Cancel upstream
               ...
            });
            
        });
        return upstream;
```

Reply:
I've tried but this doesn't seem to work😂 because `CompletionStage<HttpResponse>` will never be completed when the further downstream(e.g. `DeferredStreamMessage`) gets cancelled earlier.

Reply:
Checked and it was a wrong comment. 😅